core/http/src/uri/uri.rs

```diff
@@ -316,3 +325,264 @@ macro_rules! impl_uri_from {
+    // this method should allocate exactly one string
+    impl<'a> Serialize for Uri<'a> {
+        fn serialize<S>(&self, serializer: S) -> Result<S::Ok, S::Error>
+        where
+                S: Serializer {
+            serializer.serialize_str(&format!("{}", self))
+        }
+    }
+
+    struct UriVisitor;
+
+    impl<'de> Visitor<'de> for UriVisitor {
+        type Value = Uri<'de>;
+        fn expecting(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+            write!(f, "a URI")
+        }
+
+        fn visit_borrowed_str<E>(self, v: &'de str) -> Result<Self::Value, E>
+        where
+                E: de::Error, {
+            Uri::try_from(v).map_err(de::Error::custom)
+        }
+        fn visit_str<E>(self, v: &str) -> Result<Self::Value, E>
+        where
+                E: de::Error, {
+            Uri::try_from(v.to_string()).map_err(de::Error::custom)
+        }
+        fn visit_string<E>(self, v: String) -> Result<Self::Value, E>
+        where
+                E: de::Error, {
+            Uri::try_from(v).map_err(de::Error::custom)
+        }
+    }
+
+    impl<'de> Deserialize<'de> for Uri<'de> {
+        fn deserialize<D>(deserializer: D) -> Result<Self, D::Error>
+        where
+                D: Deserializer<'de> {
+            deserializer.deserialize_str(UriVisitor)
+        }
+    }
+    
+    #[cfg(test)]
+    mod tests {
+        use super::*;
+        
+        use std::convert::TryFrom;
+        use crate::{ext::IntoOwned, uri::{self, Origin, Authority, Absolute, Host}};
+        #[derive(Debug)]
+        enum TestError {
+            UriError(Box<uri::Error<'static>>),
+            JsonError(serde_json::Error),
+        }
+        impl<'a> From<uri::Error<'a>> for TestError {
+            fn from(e: uri::Error<'a>) -> Self {
+                Self::UriError(Box::new(e.into_owned()))
+            }
+        }
+        impl From<serde_json::Error> for TestError {
+            fn from(e: serde_json::Error) -> Self {
+                Self::JsonError(e)
+            }
+        }
+
+        // Test Helper function
+        // 'a sould almost always be 'static, but it may be useful to not require that
+        fn test_helper<'a>(raw: &str, json: &'a str, uri_obj: Uri<'a>) -> Result<(), TestError> {
+            let uri = Uri::try_from(raw).expect(raw);
+            assert_eq!(uri, uri_obj, "Uri::try_from did not create the expected Uri: {} != {}", raw, uri_obj);
+
+            assert_eq!(serde_json::to_string(&uri)?, json, "Serialization did not produce the expected Json string");
+            assert_eq!(serde_json::from_str::<Uri<'a>>(json)?, uri, "Deserialization from a `&str` did not produce the expected Uri");
+            Ok(())
+        }
+        
+        /// /path?query
+        fn origin<'a>(path: &'a str, query: Option<&'a str>) -> Uri<'a> {
+            Uri::Origin(Origin::new(path, query))
+        }
```

Comment:
Do we need this? `Uri` implements `PartialEq<Origin>`, and this is otherwise `Origin::new(path, query)`.